File: ios/AetherMeshKit/Sources/AetherMeshKit/Crypto/ChatCrypto.swift
import Crypto
import CryptoExtras
import Foundation

/// PBKDF2-HMAC-SHA256, one 32-byte block. Identical output to Android
/// `ChatKeyDerivation` and firmware `PacketAuth` (standard PBKDF2 with dkLen 32).
public enum ChatKeyDerivation {
    public static let iterations = 120_000

    public static func derive(passcode: String, salt: Data, iterations: Int = iterations) throws -> SymmetricKey {
        try KDF.Insecure.PBKDF2.deriveKey(
            from: Data(passcode.utf8),
            salt: salt,
            using: .sha256,
            outputByteCount: 32,
            // 120k matches every deployed node and app; the library's 210k floor
            // would produce keys nothing else in the mesh can use.
            unsafeUncheckedRounds: iterations
        )
    }
}

/// Chat payload crypto, wire-compatible with Android `ChatCrypto`.
///
/// v2 (only format this app writes): `"v2:" + base64(salt16 || iv12 || ciphertext || tag16)`,
/// AES-256-GCM, key from [ChatKeyDerivation] with the per-message salt, and the
/// chat context label as AAD. Decrypt also accepts v1 GCM (SHA-256 of the
/// passcode, iv12 prefix). Legacy AES-ECB is deliberately not supported: it has
/// no integrity check, so a forged message could not be told from a real one.
public enum ChatCrypto {
    public static let errorInvalid = "[Decryption Error - Invalid Message]"
    public static let errorBadContext = "[Decryption Error - Bad Key or Context]"
    public static let errorBadKey = "[Decryption Error - Bad Key]"

    static let v2Prefix = "v2:"
    static let saltLength = 16
    static let ivLength = 12
    static let tagLength = 16

    /// Returns nil on failure. Callers must refuse to send rather than fall back to plaintext.
    public static func encrypt(_ plainText: String, passcode: String, chatContext: String = "") -> String? {
        encrypt(
            plainText,
            passcode: passcode,
            chatContext: chatContext,
            salt: randomBytes(saltLength),
            iv: randomBytes(ivLength)
        )
    }

    static func encrypt(_ plainText: String, passcode: String, chatContext: String, salt: Data, iv: Data) -> String? {
        do {
            let key = try ChatKeyDerivation.derive(passcode: passcode, salt: salt)
            let nonce = try AES.GCM.Nonce(data: iv)
            let sealed = try AES.GCM.seal(
                Data(plainText.utf8),
                using: key,
                nonce: nonce,
                authenticating: Data(chatContext.utf8)
            )
            var payload = Data()
            payload.append(salt)
            payload.append(iv)
            payload.append(sealed.ciphertext)
            payload.append(sealed.tag)
            return v2Prefix + payload.base64EncodedString()
        } catch {
            return nil
        }
    }

    /// Plaintext, or one of the `error*` sentinels (the same strings Android stores).
    public static func decrypt(_ cipherText: String, passcode: String, chatContext: String = "") -> String {
        if cipherText.hasPrefix(v2Prefix) {
            // Android rejects anything not longer than salt + iv + tag (44 bytes).
            guard let decoded = Data(base64Encoded: String(cipherText.dropFirst(v2Prefix.count))),
                  decoded.count > saltLength + ivLength + tagLength
            else { return errorInvalid }
            let bytes = [UInt8](decoded)
            let salt = Data(bytes[0..<saltLength])
            let iv = Data(bytes[saltLength..<(saltLength + ivLength)])
            let body = Data(bytes[(saltLength + ivLength)...])
            guard body.count > tagLength else { return errorInvalid }
            do {
                let key = try ChatKeyDerivation.derive(passcode: passcode, salt: salt)
                let box = try AES.GCM.SealedBox(
                    nonce: AES.GCM.Nonce(data: iv),
                    ciphertext: body.prefix(body.count - tagLength),
                    tag: body.suffix(tagLength)
                )
                let plain = try AES.GCM.open(box, using: key, authenticating: Data(chatContext.utf8))
                return String(decoding: plain, as: UTF8.self)
            } catch {
                return errorBadContext
            }
        }

        // v1: AES-GCM keyed by SHA-256(passcode), iv12 || ciphertext || tag16, no AAD.
        guard let decoded = Data(base64Encoded: cipherText), decoded.count > ivLength + tagLength else {
            return errorBadKey
        }
        do {
            let key = SymmetricKey(data: SHA256.hash(data: Data(passcode.utf8)))
            let bytes = [UInt8](decoded)
            let body = Data(bytes[ivLength...])
            let box = try AES.GCM.SealedBox(
                nonce: AES.GCM.Nonce(data: Data(bytes[0..<ivLength])),
                ciphertext: body.prefix(body.count - tagLength),
                tag: body.suffix(tagLength)
            )
            return String(decoding: try AES.GCM.open(box, using: key), as: UTF8.self)
        } catch {
            return errorBadKey
        }
    }

    public static func isDecryptFailure(_ content: String) -> Bool {
        content == errorInvalid || content == errorBadContext || content == errorBadKey
    }

    static func randomBytes(_ count: Int) -> Data {
        var generator = SystemRandomNumberGenerator()
        return Data((0..<count).map { _ in UInt8.random(in: .min ... .max, using: &generator) })
    }
}

File: app/app/src/main/java/com/silentwolf75/aethermesh/data/ChatCrypto.kt
package com.silentwolf75.aethermesh.data

import android.util.Base64
import java.security.MessageDigest
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Chat payload crypto. v2 is AES-256-GCM with PBKDF2 salt + IV prepended.
 * Decrypt also accepts v1 GCM (SHA-256 key, IV prepended). There is no ECB path.
 * Callers must refuse to send when [encrypt] returns null — never fall back
 * to plaintext.
 */
object ChatCrypto {
    const val ERROR_INVALID = "[Decryption Error - Invalid Message]"
    const val ERROR_BAD_CONTEXT = "[Decryption Error - Bad Key or Context]"
    const val ERROR_BAD_KEY = "[Decryption Error - Bad Key]"

    fun encrypt(plainText: String, passcode: String, chatIdentifier: String = ""): String? {
        return try {
            val salt = ByteArray(16)
            SecureRandom().nextBytes(salt)
            val keySpec = ChatKeyDerivation.derive(passcode, salt)
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            val iv = ByteArray(12)
            SecureRandom().nextBytes(iv)
            cipher.init(Cipher.ENCRYPT_MODE, keySpec, GCMParameterSpec(128, iv))
            if (chatIdentifier.isNotEmpty()) {
                cipher.updateAAD(chatIdentifier.toByteArray(Charsets.UTF_8))
            }
            val encryptedBytes = cipher.doFinal(plainText.toByteArray(Charsets.UTF_8))
            "v2:" + Base64.encodeToString(salt + iv + encryptedBytes, Base64.NO_WRAP)
        } catch (_: Exception) {
            null
        }
    }

    fun decrypt(cipherText: String, passcode: String, chatIdentifier: String = ""): String {
        if (cipherText.startsWith("v2:")) {
            return try {
                val decoded = Base64.decode(cipherText.removePrefix("v2:"), Base64.NO_WRAP)
                if (decoded.size <= 44) return ERROR_INVALID
                val salt = decoded.copyOfRange(0, 16)
                val iv = decoded.copyOfRange(16, 28)
                val keySpec = ChatKeyDerivation.derive(passcode, salt)
                val cipher = Cipher.getInstance("AES/GCM/NoPadding")
                cipher.init(Cipher.DECRYPT_MODE, keySpec, GCMParameterSpec(128, iv))
                if (chatIdentifier.isNotEmpty()) {
                    cipher.updateAAD(chatIdentifier.toByteArray(Charsets.UTF_8))
                }
                String(cipher.doFinal(decoded, 28, decoded.size - 28), Charsets.UTF_8)
            } catch (_: Exception) {
                ERROR_BAD_CONTEXT
            }
        }

        val keySpec = deriveLegacyKey(passcode)
        try {
            val decoded = Base64.decode(cipherText, Base64.NO_WRAP)
            if (decoded.size > 28) {
                val cipher = Cipher.getInstance("AES/GCM/NoPadding")
                cipher.init(Cipher.DECRYPT_MODE, keySpec, GCMParameterSpec(128, decoded, 0, 12))
                return String(cipher.doFinal(decoded, 12, decoded.size - 12), Charsets.UTF_8)
            }
        } catch (_: Exception) {
            // fall through to the legacy format
        }
        // There is deliberately no AES-ECB fallback. ECB was only used to
        // encrypt on the project's first day and never shipped in a release,
        // and a fallback that tries ECB on anything GCM rejects would turn
        // roughly 1 in 256 corrupt messages into scrambled "plaintext" instead
        // of an error, since that is how often random bytes pass PKCS#5 padding.
        return ERROR_BAD_KEY
    }

    internal fun deriveLegacyKey(passcode: String): SecretKeySpec {
        val digest = MessageDigest.getInstance("SHA-256")
        val bytes = digest.digest(passcode.toByteArray(Charsets.UTF_8))
        return SecretKeySpec(bytes, "AES")
    }
}
